### DISCUSSION TITLE
issue: isDirty is not accurate in controlled onChange

### DISCUSSION THREAD

**Discussion by domi7777** [react-hook-form/react-hook-form#7393] [Q&A] [answered]
### Version Number

7.22.5

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-with-material-ui-forked-7qzw6?file=/src/index.js

### Steps to reproduce

Change the username to any value other than default, revert, see the output in the console, `fieldState.isDirty`/ `formState.isDirty` flag inside the `render` seems to be out of sync when accessed from `onChange` callback. Please let me know if I'm doing anything wrong.

![dirty-error](https://user-images.githubusercontent.com/5888729/147678986-f8d1d41c-549f-4310-ab3d-fddf984c927e.gif)

### Workaround
It seems to me that only `formState.dirtyFields` holds the correct state of "dirtyness"

### Expected behaviour

Ideally `formState.isDirty` & `fieldState.isDirty` should contain the correct state of the form field.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by SergeyWebDeveloper** [comment, marked as answer]
@domi7777 , you need to do this outside of `onChange`. Do you agree with this behavior?

https://codesandbox.io/s/react-hook-form-with-material-ui-forked-so213?file=/src/index.js

**Reply by domi7777** [reply]
Hi @SergeyWebDeveloper , thanks for your suggestion, however I'd like to have the `isDirty` value in the onChange. 

The reason is that it is forwarded to the parent component, like in this simple example: [https://codesandbox.io/s/react-hook-form-with-material-ui-forked-zrzl0?file=/src/index.js](https://codesandbox.io/s/react-hook-form-with-material-ui-forked-zrzl0?file=/src/index.js)

The idea is to keep track of the sub-components `dirty` states in the parent. Do you know of other/better ways to achieve this?

**Reply by SergeyWebDeveloper** [reply]
@domi7777 , you need to do it like this.
https://codesandbox.io/s/react-hook-form-with-material-ui-forked-qz8t3?file=/src/index.js

Or you can do it via context - https://react-hook-form.com/api/useformcontext

**Reply by domi7777** [reply]
Thanks @SergeyWebDeveloper, these are interesting ways to do it without `onChange`, I'll look into that.

If I understand correctly, one cannot rely on these properties `formState.isDirty` & `fieldState.isDirty` inside the `onChange`.

Out of curiosity, would you advise against using `formState.dirtyFields` as a workaround inside the `onChange` ?

**Reply by SergeyWebDeveloper** [reply]
@domi7777 , it is better not to do this, there should be one source of truth - `useFormState`. `onChange` is needed primarily to change the value, it's not a good idea to use the dirty field sign in it)

**Reply by bluebill1049** [reply]
@SergeyWebDeveloper thanks for helping and you are absolutely right!

**Reply by SergeyWebDeveloper** [reply]
Thank you, @bluebill1049 )

**Comment by bluebill1049** [comment]
thanks @SergeyWebDeveloper that's correct. formState will be flushed in the next render
